README.md
force-bind-seccomp
==================

Use seccomp to foce processes to bind to specific address instead of binding to
`0.0.0.0` (IPv4) or `[::]` (IPv6). It can also transform a process that does not
make use of systemd socket activation to use passed file descriptor instead of
binding their own sockets.

Under the hood, it works in two ways:

- if systemd socket activation is not needed, then it uses seccomp only. When a
  bind system call is detected, the process is stopped and if the bind address
  matches a force-bind rule, then the process memory is altered to replace the
  address given to `bind()` with a replacement address.

  It uses `SECCOMP_RET_USER_NOTIF` which is only available on recent kernels.

- if systemd socket activation is needed, then it uses seccomp in combinaison
  with ptrace and when a `bind()` system call is detected, then the process is
  stopped and ptrace is used to alter the process. The system call registers are
  dumped and if the address bound matches a pattern:

    - either ptrace is used to replace the address with a replacement address,
      just like with seccomp, and the bind system call continues

    - if systemd socket activation is needed for that pattern, then the system
      call is replaced by the `dup2()` system call and the return value is
      altered to return `0` in case of success.

In the future, when `SECCOMP_NOTIFY_IOCTL_ADDFD` will become available, then
ptrace could be entirely replaced by seccomp, including when systemd socket
activation is needed.

This is still a young project. Don't hesitate to report bugs or submit fixes.

Known bugs
----------

- parent process always exits with status 0 in seccomp-only mode
- if the process close a systemd activated socket and opens a new socket on the
  same file descriptor number that is not catched by force-bind, then the
  subsequent listen() calls will be skipped
- security issue: race condition when replacing the network address causing a
  malicious program to bind an otherwise forbidden address.
- When the file descriptors are not passed by systemd (the service is started
  while the socket was not active for example), force-bind should let the
  process bind() and listen() normally

History
-------

This project was bord out of dissatisfaction on how linux containers (docker or
podman) handled IPv6. I came to the conclusion that to work properly, the
containers needed to use the same network namespaces as the host, and they
should use the public address for communication with the Internet, and bind
their services on a private address space (127.0.0.0/8 on IPv4, ULA prefix on
IPv6) for container intercommunication.

I needed a way to force containers to bind services on those private addresses
while still being able to access the public addresses to connect to the
Internet. That's how `force-bind` was born.

I used a seccomp example from Michael Kerrisk written 2019 as a starting point
because I needed something that worked with `SECCOMP_RET_USER_NOTIF` which was
quite bleeding-edge at that time. Not much documentation was available. That's
the reason behind the GPLv3 license (that and the fact that I pretty much agree
with the FSF philosophy).

Next, I needed a way to force services (this time not in containers) that didn't
play well with systemd socket activation to use the socket file descriptor
passed by systemd and new options to `force-bind` were developed. Using proxies
didn't work well because you lose the source address in the process. Also, as
another indirection layer, it participates to system bloat.

Usage
-----

```
Usage: ./force-bind [options] TARGET_PROGRAM [ARGS ...]
Options
    -h                    Help
    -V                    Version information
    -m MATCH=ADDR         Replace bind() matching first MATCH with ADDR
    -b ADDR               Replace all bind() with ADDR (if same family)
    -d                    Deny all bind()
    -p                    Force seccomp-ptrace instead of only seccomp
    -D                    Debug messages
    -v                    Verbose
    -q                    Quiet

Last rules (-m, -b, -d) are applied first so you can put your general policy
first on your command-line and any following argument can override it.

With the -m rule, MATCH pattern can be:

    :PORT                 Matches both IPv4 and IPv6 listening on PORT
    ADDR:PORT/PREFIX      Matches address with given port with the netmark
                          corresponding to PREFIX applied first on both IPs
                          IPv6 address must be enclosed in squared brackets.

In the rules, ADDR can be:

    IP:PORT               Changes the bind target to specified address
    fd=N or fd-N          dup2() the specified inherited file descriptor
                          in place
    sd=N or sd-N          same as fd=N but for systemd file descriptor. sd=0
                          is thus equivalent to fd=3

Examples:

  * force-bind -m 0.0.0.0:80/0=127.0.0.1:8080 progname args...

    Replace binds from any IPv4 address port 80 to localhost-only port
    8080. IPv6 binds are allowed.

  * force-bind -d -m '[::]:80/0=[::1]:8080' progname args...

    Replace binds from any IPv6 address port 80 to localhost-only port
    8080. IPv4 binds are denied.

  * force-bind -d -b '[::1]:8081' -b '127.0.0.1:8080' progname args...

    Replace binds from any IPv6 address to localhost port 8081 and from
    any IPv4 address to localhost port 8080.

  * force-bind -m ':80=sd=0' -m ':80=fd=4' progname args...

    Replace any bind to port 80 using first socket from systemd socket
    activation. Any port 81 is replaced by passed file descriptor 4.


Version: v0.0.0-19-g2207165-dirty
```

Test
----

    make force-bind target-bind parent-socket-activate
    ./force-bind -v -b 127.0.0.1:16379 ./target-bind 7777
    nc -u 127.0.0.1 16379

    ./parent-socket-activate 127.0.0.1 8888 ./force-bind -v -b 0.0.0.0:16379/0=sd=0 ./target-bind 7777
    nc -u 127.0.0.1 8888


Reference
---------

In random order:

- https://people.kernel.org/brauner/the-seccomp-notifier-new-frontiers-in-unprivileged-container-development
- https://nullprogram.com/blog/2018/06/23/
- https://github.com/alfonsosanchezbeato/ptrace-redirect
- https://github.com/emptymonkey/ptrace_do
- http://www.alfonsobeato.net/c/filter-and-modify-system-calls-with-seccomp-and-ptrace/
- http://www.alfonsobeato.net/c/modifying-system-call-arguments-with-ptrace/
- http://man7.org/tlpi/code/online/dist/seccomp/seccomp_user_notification.c
- https://news.ycombinator.com/item?id=19187417
- https://2019.linux.conf.au/schedule/presentation/236/ (Tue 22 Jan 2019)
- https://youtube.com/watch?v=sqvF_Mdtzgg
- https://github.com/Intika-Linux-Firewall/Force-Bind
- http://kernel.embedromix.ro/us/ http://kernel.embedromix.ro/us/force_bind/

[x86_64 syscall registers](http://blog.rchapman.org/posts/Linux_System_Call_Table_for_x86_64/) ([and how to use it](https://stackoverflow.com/questions/33431994/extracting-system-call-name-and-arguments-using-ptrace):

- regs.rdi - Stores the first argument
- regs.rsi - Stores the second argument
- regs.rdx - Stores the third argument
- regs.r10 - Stores the fourth argument
- regs.r8 - Stores the fifth argument
- regs.r9 - Stores the sixth argument



CHANGELOG.md
Changelog
=========

Bug fixes:

- tracer process exits with child status (seccomp-ptrace mode only)
- listen call is disabled for systemd activated sockets

v0.1.0
------

First released version of force-bind.


ip_funcs.h
#ifndef _FORCE_BIND_SECCOMP_IP_FUNCS_H_
#define _FORCE_BIND_SECCOMP_IP_FUNCS_H_

#include <sys/types.h>
#include <sys/socket.h>
#include <netdb.h>


static int getaddrinfo2(const char *nodeport0,
                       const struct addrinfo *hints,
                       struct addrinfo **res) {
    size_t len = strlen(nodeport0);
    char nodeport[len+1];
    strncpy(nodeport, nodeport0, len+1);

    char *node = nodeport, *service = nodeport;

    if(*nodeport == '['){
        char *c = strchr(nodeport, ']');
        if(c && *c) {
            *c = 0;
            node = nodeport + 1;
            service = c + 1;
        }
    }

    char *c = strchr(service, ':');
    if(c && *c) {
        *c = 0;
        service = c+1;
    } else {
        service = NULL;
    }

    return getaddrinfo(node, service, hints, res);
}

static char *
get_ip_str(const struct sockaddr *sa, char *s, size_t maxlen)
{
#define sa4 ((struct sockaddr_in *)sa)
#define sa6 ((struct sockaddr_in6 *)sa)
    char addr[1024];

    if(!sa) {
        strncpy(s, "(null)", maxlen);
        return s;
    }

    switch(sa->sa_family) {
        case AF_INET: {
            inet_ntop(AF_INET, &(sa4->sin_addr), addr, sizeof(addr));
            snprintf(s, maxlen, "%s:%d", addr, ntohs(sa4->sin_port));
            break;
        }

        case AF_INET6:
            inet_ntop(AF_INET6, &(sa6->sin6_addr), addr, sizeof(addr));
            snprintf(s, maxlen, "[%s]:%d", addr, ntohs(sa6->sin6_port));
            break;

        default:
            return NULL;
    }

    return s;
#undef sa4
#undef sa6
}

#endif


target_mkdir.c
#define _GNU_SOURCE
#include <sys/types.h>
#include <sys/prctl.h>
#include <fcntl.h>
#include <limits.h>
#include <signal.h>
#include <sys/wait.h>
#include <stddef.h>
#include <stdbool.h>
#include <linux/audit.h>
#include <sys/syscall.h>
#include <sys/stat.h>
#include <linux/filter.h>
#include <linux/seccomp.h>
#include <sys/ioctl.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>

int main(int argc, char **argv) {
    int s;

    for (char **ap = argv+1; *ap != NULL; ap++) {
        printf("\nTarget process: about to make directory \"%s\"\n", *ap);
        s = mkdir(*ap, 0600);
        if (s == -1)
            perror("Target process: mkdir");
        else
            printf("Target process: SUCCESS: mkdir(2) returned = %d\n", s);
    }

    printf("Target process: terminating\n");
    exit(EXIT_SUCCESS);
}


scm_functions.c
/*************************************************************************\
*                  Copyright (C) Michael Kerrisk, 2019.                   *
*                                                                         *
* This program is free software. You may use, modify, and redistribute it *
* under the terms of the GNU Lesser General Public License as published   *
* by the Free Software Foundation, either version 3 or (at your option)   *
* any later version. This program is distributed without any warranty.    *
* See the files COPYING.lgpl-v3 and COPYING.gpl-v3 for details.           *
\*************************************************************************/

/* scm_functions.c

   Functions to exchange ancillary data over UNIX domain sockets.
   These functions are simplistic, in that they ignore the "real" data
   content on the assumption that the sockets are being used only for
   the purposes of exchanging ancillary data. In many real-world
   applications, the application makes use of both the "real" data
   channel and the ancillary data, with some kind of protocol that
   determines how the "real" and ancillary data are used together.
*/
#include "scm_functions.h"

/* Send the file descriptor 'fd' over the connected UNIX domain socket
   'sockfd'. Returns 0 on success, or -1 on error. */

int
sendfd(int sockfd, int fd)
{
    struct msghdr msgh;
    struct iovec iov;
    int data;
    struct cmsghdr *cmsgp;

    /* Allocate a char array of suitable size to hold the ancillary data.
       However, since this buffer is in reality a 'struct cmsghdr', use a
       union to ensure that it is aligned as required for that structure.
       Alternatively, we could allocate the buffer using malloc(), which
       returns a buffer that satisfies the strictest alignment requirements
       of any type. However, if we employ that approach, we must ensure
       that we free() the buffer on all return paths from this function. */
    union {
        char   buf[CMSG_SPACE(sizeof(int))];
                        /* Space large enough to hold an 'int' */
        struct cmsghdr align;
    } controlMsg;

    /* The 'msg_name' field can be used to specify the address of the
       destination socket when sending a datagram. However, we do not
       need to use this field because we presume that 'sockfd' is a
       connected socket. */

    msgh.msg_name = NULL;
    msgh.msg_namelen = 0;

    /* On Linux, we must transmit at least one byte of real data in
       order to send ancillary data. We transmit an arbitrary integer
       whose value is ignored by recvfd(). */

    msgh.msg_iov = &iov;
    msgh.msg_iovlen = 1;
    iov.iov_base = &data;
    iov.iov_len = sizeof(int);
    data = 12345;

    /* Set 'msghdr' fields that describe ancillary data */

    msgh.msg_control = controlMsg.buf;
    msgh.msg_controllen = sizeof(controlMsg.buf);

    /* Set up ancillary data describing file descriptor to send */

    cmsgp = CMSG_FIRSTHDR(&msgh);
    cmsgp->cmsg_level = SOL_SOCKET;
    cmsgp->cmsg_type = SCM_RIGHTS;
    cmsgp->cmsg_len = CMSG_LEN(sizeof(int));
    *((int *) CMSG_DATA(cmsgp)) = fd;

    /* Send real plus ancillary data */

    if (sendmsg(sockfd, &msgh, 0) == -1)
        return -1;

    return 0;
}

/* Receive a file descriptor on a connected UNIX domain socket. Returns
   the received file descriptor on success, or -1 on error. */

int
recvfd(int sockfd)
{
    struct msghdr msgh;
    struct iovec iov;
    int data;
    ssize_t nr;

    /* Allocate a char buffer for the ancillary data. See the comments
       in sendfd() */
    union {
        char   buf[CMSG_SPACE(sizeof(int))];
        struct cmsghdr align;
    } controlMsg;
    struct cmsghdr *cmsgp;

    /* The 'msg_name' field can be used to obtain the address of the
       sending socket. However, we do not need this information. */

    msgh.msg_name = NULL;
    msgh.msg_namelen = 0;

    /* Specify buffer for receiving real data */

    msgh.msg_iov = &iov;
    msgh.msg_iovlen = 1;
    iov.iov_base = &data;       /* Real data is an 'int' */
    iov.iov_len = sizeof(int);

    /* Set 'msghdr' fields that describe ancillary data */

    msgh.msg_control = controlMsg.buf;
    msgh.msg_controllen = sizeof(controlMsg.buf);

    /* Receive real plus ancillary data; content of real data is ignored */

    nr = recvmsg(sockfd, &msgh, 0);
    if (nr == -1)
        return -1;

    cmsgp = CMSG_FIRSTHDR(&msgh);

    /* Check the validity of the 'cmsghdr' */

    if (cmsgp == NULL ||
            cmsgp->cmsg_len != CMSG_LEN(sizeof(int)) ||
            cmsgp->cmsg_level != SOL_SOCKET ||
            cmsgp->cmsg_type != SCM_RIGHTS) {
        errno = EINVAL;
        return -1;
    }

    /* Return the received file descriptor to our caller */

    return *((int *) CMSG_DATA(cmsgp));
}


scm_functions.h
/*************************************************************************\
*                  Copyright (C) Michael Kerrisk, 2019.                   *
*                                                                         *
* This program is free software. You may use, modify, and redistribute it *
* under the terms of the GNU Lesser General Public License as published   *
* by the Free Software Foundation, either version 3 or (at your option)   *
* any later version. This program is distributed without any warranty.    *
* See the files COPYING.lgpl-v3 and COPYING.gpl-v3 for details.           *
\*************************************************************************/

/* scm_functions.h

   Functions to exchange ancillary data over a UNIX domain socket.
*/
#ifndef SCM_FUNCTIONS_H
#define SCM_FUNCTIONS_H         /* Prevent accidental double inclusion */

#include <sys/socket.h>
#include <sys/un.h>
#include <stdio.h>
#include <errno.h>

int sendfd(int sockfd, int fd);

int recvfd(int sockfd);

#endif


parent_soocket_activate.c
#include <sys/types.h>
#include <sys/wait.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <sys/socket.h>
#include <netdb.h>
#include <arpa/inet.h>

#include "ip_funcs.h"

#define BUF_SIZE 500

#define errExit(msg)    do { perror(msg); exit(EXIT_FAILURE); \
                        } while (0)

int main(int argc, char *argv[]) {
   char addrstr[1024];
   struct addrinfo hints;
   struct addrinfo *result, *rp;
   int sfd, s;
   struct sockaddr_storage peer_addr;
   socklen_t peer_addr_len;
   ssize_t nread;
   char buf[BUF_SIZE];

   if (argc < 4) {
       fprintf(stderr, "Usage: %s addr port\n", argv[0]);
       exit(EXIT_FAILURE);
   }

   memset(&hints, 0, sizeof(struct addrinfo));
   hints.ai_family = AF_UNSPEC;    /* Allow IPv4 or IPv6 */
   hints.ai_socktype = SOCK_DGRAM; /* Datagram socket */
   hints.ai_flags = AI_PASSIVE;    /* For wildcard IP address */
   hints.ai_protocol = 0;          /* Any protocol */
   hints.ai_canonname = NULL;
   hints.ai_addr = NULL;
   hints.ai_next = NULL;

   s = getaddrinfo(argv[1], argv[2], &hints, &result);
   if (s != 0) {
       fprintf(stderr, "getaddrinfo: %s\n", gai_strerror(s));
       exit(EXIT_FAILURE);
   }

   /* getaddrinfo() returns a list of address structures.
      Try each address until we successfully bind(2).
      If socket(2) (or bind(2)) fails, we (close the socket
      and) try the next address. */

   for (rp = result; rp != NULL; rp = rp->ai_next) {
       printf("parent: socket(%d, %d, %d)\n", rp->ai_family, rp->ai_socktype, rp->ai_protocol);
       sfd = socket(rp->ai_family, rp->ai_socktype,
               rp->ai_protocol);
       if (sfd == -1)
           continue;

       printf("parent: bind(%d, %p, %d)\n", sfd, (void*) rp->ai_addr, rp->ai_addrlen);
       printf("parent (before): %p = %s\n", rp->ai_addr, get_ip_str(rp->ai_addr, addrstr, sizeof(addrstr)));
       if (bind(sfd, rp->ai_addr, rp->ai_addrlen) == 0) {
           printf("parent (after):  %p = %s\n", rp->ai_addr, get_ip_str(rp->ai_addr, addrstr, sizeof(addrstr)));
           break;                  /* Success */
       }
       printf("parent (after):  %p = %s\n", rp->ai_addr, get_ip_str(rp->ai_addr, addrstr, sizeof(addrstr)));

       close(sfd);
   }

   if (rp == NULL) {               /* No address succeeded */
       perror("Could not bind");
       exit(EXIT_FAILURE);
   }

   if(sfd != 3) {
      dup2(3, sfd);
   }

   pid_t child = fork();
   if (child == 0) {
      char pid_str[64];
      snprintf(pid_str, 64, "%d", getpid());

      setenv("LISTEN_FDS", "1", 1);
      setenv("LISTEN_PID", pid_str, 1);

      char **args = &argv[3];
      if(execvp(args[0], args) == -1)
         errExit("execvp");
      exit(EXIT_FAILURE);
   }

   waitpid(child, 0, 0);
   exit(EXIT_SUCCESS);
}


target_bind.c
#include <sys/types.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <sys/socket.h>
#include <netdb.h>
#include <arpa/inet.h>

#include "ip_funcs.h"

#define BUF_SIZE 500

int main(int argc, char *argv[]) {
   char addrstr[1024];
   struct addrinfo hints;
   struct addrinfo *result, *rp;
   int sfd, s;
   struct sockaddr_storage peer_addr;
   socklen_t peer_addr_len;
   ssize_t nread;
   char buf[BUF_SIZE];

   if (argc != 2) {
       fprintf(stderr, "Usage: %s port\n", argv[0]);
       exit(EXIT_FAILURE);
   }

   memset(&hints, 0, sizeof(struct addrinfo));
   hints.ai_family = AF_UNSPEC;    /* Allow IPv4 or IPv6 */
   hints.ai_socktype = SOCK_DGRAM; /* Datagram socket */
   hints.ai_flags = AI_PASSIVE;    /* For wildcard IP address */
   hints.ai_protocol = 0;          /* Any protocol */
   hints.ai_canonname = NULL;
   hints.ai_addr = NULL;
   hints.ai_next = NULL;

   s = getaddrinfo(NULL, argv[1], &hints, &result);
   if (s != 0) {
       fprintf(stderr, "getaddrinfo: %s\n", gai_strerror(s));
       exit(EXIT_FAILURE);
   }

   /* getaddrinfo() returns a list of address structures.
      Try each address until we successfully bind(2).
      If socket(2) (or bind(2)) fails, we (close the socket
      and) try the next address. */

   for (rp = result; rp != NULL; rp = rp->ai_next) {
       sfd = socket(rp->ai_family, rp->ai_socktype,
               rp->ai_protocol);
       printf("target: socket(%d, %d, %d) = %d\n", rp->ai_family, rp->ai_socktype, rp->ai_protocol, sfd);
       if (sfd == -1)
           continue;

       printf("target: bind(%d, %p, %d)\n", sfd, (void*) rp->ai_addr, rp->ai_addrlen);
       printf("target (before): %p = %s\n", rp->ai_addr, get_ip_str(rp->ai_addr, addrstr, sizeof(addrstr)));
       int res =bind(sfd, rp->ai_addr, rp->ai_addrlen);
       if (res == 0) {
           printf("target (after):  %p = %s\n", rp->ai_addr, get_ip_str(rp->ai_addr, addrstr, sizeof(addrstr)));
           break;                  /* Success */
       }
       printf("target: bind() returned with %d\n", res);
       perror("target");
       printf("target (after):  %p = %s\n", rp->ai_addr, get_ip_str(rp->ai_addr, addrstr, sizeof(addrstr)));

       close(sfd);
   }

   if (rp == NULL) {               /* No address succeeded */
       perror("target: Could not bind");
       exit(EXIT_FAILURE);
   }

   freeaddrinfo(result);           /* No longer needed */

   /* Read datagrams and echo them back to sender */

   for (;;) {
       printf("target: recvfrom(%d, %p, %d, ...)\n", sfd, buf, BUF_SIZE);
       peer_addr_len = sizeof(struct sockaddr_storage);
       nread = recvfrom(sfd, buf, BUF_SIZE, 0,
               (struct sockaddr *) &peer_addr, &peer_addr_len);
       if (nread == -1)
           continue;               /* Ignore failed request */

       char host[NI_MAXHOST], service[NI_MAXSERV];

       s = getnameinfo((struct sockaddr *) &peer_addr,
                       peer_addr_len, host, NI_MAXHOST,
                       service, NI_MAXSERV, NI_NUMERICSERV);
       if (s == 0)
           printf("target: Received %zd bytes from %s:%s\n",
                   nread, host, service);
       else
           fprintf(stderr, "getnameinfo: %s\n", gai_strerror(s));

       if (sendto(sfd, buf, nread, 0,
                   (struct sockaddr *) &peer_addr,
                   peer_addr_len) != nread)
           fprintf(stderr, "target: Error sending response\n");
   }
}
